algorithms/mopt/mopt_hierarflow_routines.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_hierarflow_routines.hpp"
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option_get_splice_cycles.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::algorithm::mopt::routine {
+
+namespace other {
+
+using mopt::option::MOptTag;
+
+MOptUpdate::MOptUpdate(MOptState &state) : state(state) {}
+
+MOptMidCalleeRef MOptUpdate::operator()(
+    [[maybe_unused]] std::shared_ptr<MOptTestcase> testcase) {
+  // retrive optimizer values
+  auto last_splice_cycle = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::LastSpliceCycle, true);
+  auto havoc_operator_finds = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::HavocOperatorFinds, true);
+
+  auto &mopt = state.mopt;
+
+  // update havoc_operator_finds
+  for (size_t i = 0; i < havoc_operator_finds.size(); i++) {
+    mopt->havoc_operator_finds[state.core_mode ? 1 : 0][i] +=
+        havoc_operator_finds[i];
+  }
+
+  if (last_splice_cycle >= afl::option::GetSpliceCycles(state)) {
+    state.UpdateSpliceCycles();
+  }
+
+  if (state.pacemaker_mode) {
+    // TODO
```

Comment:
what kind of TODO?

Reply:
fixed in 2856b4dc4eff8de57f08c750de5945b8156dc17b
forgot to implement algorithm to exit pacemaker mode